lib/translator/translation.cc

```diff
@@ -0,0 +1,73 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "common.h"
+
+namespace translator {
+
+StatusCode Translation::ScsiToNvme(absl::Span<const uint8_t> scsi_cmd,
+                                   const ScsiContext& scsi_context) {
+  this->pipeline_status = StatusCode::kSuccess;
+  // Verify buffer is large enough to contain opcode (one byte)
+  if (scsi_cmd.size() < 1) {
+    DebugLog("Empty SCSI Buffer");
+    this->pipeline_status = StatusCode::kFailure;
+    return StatusCode::kFailure;
+  }
+
+  this->scsi_cmd = scsi_cmd;
+  this->scsi_context = scsi_context;
+
+  scsi_defs::OpCode opc = static_cast<scsi_defs::OpCode>(scsi_cmd[0]);
+  switch (opc) {
+    case scsi_defs::OpCode::kInquiry:
+      return StatusCode::kSuccess;
+    default:
+      DebugLog("Bad OpCode");
+      this->pipeline_status = StatusCode::kFailure;
```

Comment:
I'm not too sure here when/how the pipeline_status differs from the StatusCode that's returned by the function.

Reply:
Pipeline status allows the translation library to keep track of errors that have occurred in upstream interactions and act accordingly without breaking the interface.

For example, it could be the case that there was no translation possible for some given scsi_data. The translation object will set its internal state to NoTranslation. As the pipeline continues, consumers get the NVMe commands using GetNvmeCmds (which returns a span of size 0), and then consumers proceed to call NvmeToScsi.
NvmeToScsi will recognize that the earlier translation attempt failed and can then respond with appropriate error codes. Meanwhile, the consumer does not notice any change in the library interface.

Reply:
It could be useful to define an enum private to the Translation class purposed for just maintaining state aswell